### PR TITLE
feat: Adding ability to specify imagePullSecrets for pipelines

### PR DESCRIPTION
**Pull request by domderen**
This PR solves #183. I didn't add any tests or examples using it or docs. What do you think we should add for it? I tested it locally to confirm it's working.

### CODE REVIEW

**Comment by alexec** [comment]
I think we want this, but my question is - would you want the same secret for all pipelines? Or do you need it for each pipeline? If it the they former, we could have an env var to define it.

**Comment by domderen** [comment]
Hmm that's a good question. I tried to mimic what Argo-Workflows is doing, and as far as I can see, out there it is done on a per workflow basis. So I thought the same might be applicable here.

Having said that, in my particular use-case, I would just need one imagePullSecret that I could use globally in Argo-Dataflow. So in my case it would make sense to have an environment variable that allows doing it. But to be honest, I'm not sure if I would go that way, or if I would just provide imagePullSecrets on my pipelines directly... I think doing the latter is more explicit, telling everyone who would like to use such a pipeline, that it requires a special secret. 

To sum up, I'm happy to provide an environment variable for it, as it might be helpful to have it as a global setting, but I would still prefer to use it on Pipeline/Step objects.

**Comment by alexec** [comment]
Let’s go with a default value for MVP. We can always add one later.

**Comment by domderen** [comment]
I'm not sure what you mean by that...

**Comment by domderen** [comment]
How's this for ability to specify ImagePullSecrets as:
- Environment variable,
- Property on a Pipeline Spec,
- Property on a Step Spec,

At the moment each of those settings defines an absolute value of imgePullSecrets. So if imagePullSecrets is defined on a Step Spec, it will override any value set in Pipeline Spec or env var. I'm wondering if it should extend previous values, rather than override them?

So if env var is set to `secret1,secret2`, PipelineSpec.ImagePullSecrets is set to `[{Name: "secret3"}]` and StepSpec.ImagePullSecrets is set to `[{Name: "secret4"}]`, the final pod spec will use just `[{Name: "secret4"}]`. I'm wondering, if instead it should merge all of those values and use `[{Name: "secret1"}, {Name: "secret2"}, {Name: "secret3"}, {Name: "secret4"}]`?

**Comment by alexec** [comment]
I guess my question is - do you think people will use the ability to specify it at pipeline or step level?

**Comment by domderen** [comment]
Hmm, I'm not sure... You have more experience with it through other Argo projects, so if you think there's no need for it, I can remove it.

But my thinking was, that if someone is running Argo-Dataflow as a cluster-wide service, and sysadmins are responsible for maintaining it, while developers are actual users, then developers don't need sysadmins to specify ImagePullSecrets. I just like that freedom...

**Comment by alexec** [comment]
That seems reasonable. How about then just specifying it on the pipeline, not the step?

**Comment by domderen** [comment]
Sure, I'll make a change. How about overriding values vs merging them? What do you think should happen if both `ARGO_DATAFLOW_IMAGE_PULL_SECRETS` env var & `PipelineSpec.imagePullSecrets` are specified? Should the second one override the first one, or merge with it?

**Comment by domderen** [comment]
Hmm, actually... I think I would still need to pass that value from `PipelineSpec` to `StepSpec` somehow... I'm looking at the `Reconcile` function in `step_controller.go` and I don't see any reference to PipelineSpec there. So to pass the value from PipelineSpec to PodSpec, I think I would need to pass it via the StepSpec anyway, right?

Or am I missing something here?

**Comment by alexec** [comment]
Good point.Pipelines don't know about pods. In that case, only the step should have the field.

**Comment by domderen** [comment]
Ok, ImagePullSecrets can now be specified as:
- Environment variable `ARGO_DATAFLOW_IMAGE_PULL_SECRETS` on the dataflow-controller,
- as `imagePullSecrets` in StepSpec,

If it is defined in StepSpec, it will override anything set by the env var.

**Review by alexec — APPROVED** [review]

**Comment by alexec** [comment]
LGTM!
